fb3370 V2014.07

andreamerello opened the pull request:
Hi Daniel,

I've also fixed the 'hang-after-a-while', (that was a switch-off-after-a-while indeed) bug.

So, let me recap:
- I've ported fb3370 on v2014.07 - we have also DDR tune (even if probably we don't need id).
- u-boot stays running without hangs (or similar issues) - both SPI and UART boot
- I've lit the power LED, so that the device doesn't seem dead :)
- ETH is working on all PHYs (I've also changed phy FW version for my device, but this is not included in this PR, because I've assumed this is not the same for every device)
- I can use both the kernel.org toolchain (gcc 4.9) as well as the openwrt toolchain (gcc 5).. (I don't know if you want this, but eventually dropping these commits from the PR should be trivial)

Remaining issues:
- NAND is still not solved 
- minor issue: the flash SPI detection message appares two times.

So I've collected all the stuff above and tidied them up a bit :)

danielschwierzeck commented on board/avm/fb3370/fb3370.c:
```diff
@@ -0,0 +1,144 @@
+#define do_gpio_init	1
+#define do_pll_init	0
+#define do_dcdc_init	1
+#elif defined(CONFIG_SYS_BOOT_NOR)
```
you can drop the branch for CONFIG_SYS_BOOT_NOR

danielschwierzeck commented on board/avm/fb3370/fb3370.c:
```diff
@@ -0,0 +1,144 @@
+	const enum ltq_gphy_clk clk = LTQ_GPHY_CLK_25MHZ_PLL0;
+	const ulong fw_addr = 0x80FF0000;
+
+	ltq_gphy_phy11g_a1x_load(fw_addr);
```
you could use ltq_soc_is_vrx200_v2() to load either the v2 firmware or the v1 firmware

danielschwierzeck commented on include/configs/fb3370.h:
```diff
@@ -0,0 +1,120 @@
+#define CONFIG_SYS_NAND_PAGE_SIZE	2048
+#define CONFIG_SYS_NAND_OOBSIZE		64
+#define CONFIG_SYS_NAND_BLOCK_SIZE	(256 * 1024)
+#define CONFIG_SYS_NAND_5_ADDR_CYCLE
```
you can drop this section because you have no NAND boot

danielschwierzeck commented on include/configs/fb3370.h:
```diff
@@ -0,0 +1,120 @@
+#define CONFIG_ENV_OFFSET_REDUND	(CONFIG_ENV_OFFSET + CONFIG_ENV_SECT_SIZE)
+
+#define MTDPARTS_DEFAULT		\
+	"mtdparts=nand-xway:256k(uboot),128k(uboot_env),128k(uboot_env_redund)"
```
you can drop the branch for CONFIG_SYS_BOOT_NANDSPL

danielschwierzeck commented on include/configs/fb3370.h:
```diff
@@ -0,0 +1,120 @@
+#define CONFIG_ENV_SECT_SIZE		(64 * 1024)
+
+#define MTDPARTS_DEFAULT		\
+	"mtdparts=spi0.4:256k(uboot),256k(uboot_env)"
```
you should change that to "mtdparts=spi0.4:192k(uboot),64k(uboot_env)" to match the dimensions of the SPI flash

danielschwierzeck commented on include/configs/fb3370.h:
```diff
@@ -0,0 +1,120 @@
+/*#define CONFIG_FLASH_CFI_MTD */
+#define CONFIG_CMD_MTD
+#define CONFIG_CMD_MTDPARTS
+#define MTDIDS_DEFAULT			"nor0=cfi,nor1=spi0.4,nand0=nand-xway"
```
drop the "nor0=cfi" part

andreamerello commented:
Ok, will fix all you suggested.
In the mean time, I've added a commit for detecting SoC version and uploading right PHY FW

andreamerello commented:
Done :)

I've done and kept further commits with those fixes in another branch, and I assumed you wanted a rebase and force-commit here.. Hope this is ok :)

danielschwierzeck commented:
Thanks, looks good now. But I think I'll cherry-pick your patches directly because I have to tidy up the openwrt/v2014.07 branch at first.

andreamerello commented:
Hi Daniel,

Hope you had a great xmas break :)

I've updated the PR with fix for NAND flash not being detected and with fix for saveenv that self-destroyed u-boot.

The NAND chip now is now recognized (I discovered that I have a 128Mb flash while openwrt pages say this device has 512Mb chip.. I don't really know if there are actually two versions of fb3370).

The problem was about the CS as you suggested; I was convinced that CS1 was routed to the NAND chip, because it actually IS.. But it is routed to the #CS2 pin of the chip instead of #CS1 (they are so close that when I checked with DMM I must have done a mistake). However the #CS2 pin is indeed not used for this specific flash chip..

So I changed u-boot to use CS0, as you initially suggested, but I've had to hack a bit more than I did in my first try to make it working.

My understanding of the EBU/NAND stuff is that:
- the EBU regions are in relation with CS pins (region0 for CS0 pin, region1 for CS1 pin)
- the NAND controller flags NAND_CON_OUT_CS1 and NAND_CON_IN_CS1 are for 'binding' the NAND controller to EBU region 1 instead of 0.

Is this correct ?

BTW I've successfully wrote the flash kernel partition from linux (tftp-booted) and u-boot is now able to retrieve the kernel from the NAND and to boot (FS on USB device for now, and still a bit of hacks, but seems to work). 

I didn't touch the linux xway NAND driver neither the DTS (thery are from LEDE git repo). Also, I've trusted (hoped) that wait cycles and other u-boot EBU configs are OK.. 

However: 
- I couldn't retrieve anything from the NAND before I flashed it, due to storm of ECC errors. I don't know if it was really invalid, or if we are using a different ECC than the original one.
- Linux complains with this mesage during boot:
   [    0.948889] nand: WARNING: 14000000.nand-parts: the ECC used on your system is too weak compared to the one required by the NAND chip

So maybe the type of ECC is really not the correct one?

About the SF/environment: I've adjusted offsets so that the env block fits without overwriting anything. I've also enabled 4K sector erase for this SF chip. I was surprised that it was not enabled even in latest u-boot git tree, but the SF datasheet claims that the chip supports it, and it seem working.. Am I missing something ?

Andrea

danielschwierzeck commented:
I've cleaned up and updated openwrt/v2014.07 with all bugfixes and features I had. I also cherry-picked the gcc5 pacthes from mainline. The NAND and VRX200 EBU driver is now also fixed for multiple chip-selects. Please rebase and try again. 

Note: you need to add "#define CONFIG_LTQ_NAND_CS0" to "include/configs/fb3370.h" in order to switch the drivers to CS0.

andreamerello commented:
On Tue, Jan 3, 2017 at 3:12 AM, Daniel Schwierzeck <notifications@github.com
> wrote:

> I've cleaned up and updated openwrt/v2014.07 with all bugfixes and
> features I had. I also cherry-picked the gcc5 pacthes from mainline. The
> NAND and VRX200 EBU driver is now also fixed for multiple chip-selects.
> Please rebase and try again.
>
> Note: you need to add "#define CONFIG_LTQ_NAND_CS0" to
> "include/configs/fb3370.h" in order to switch the drivers to CS0.
>

Awesome! I'm going to do this in next days.

Thank you very much
Andrea


> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danielschwierzeck/u-boot-lantiq/pull/17#issuecomment-270042323>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH_bMqv1e56iFQbRitg9_vhGRaYxswGrks5rOa55gaJpZM4LUSuu>
> .
>
